dbms/src/DataStreams/WindowBlockInputStream.h

```diff
@@ -0,0 +1,254 @@
+    std::vector<size_t> order_column_indices;
+
+    // Per-window-function scratch spaces.
+    std::vector<WindowFunctionWorkspace> workspaces;
+
+    std::unique_ptr<Arena> arena;
+
+    // A sliding window of blocks we currently need. We add the input blocks as
+    // they arrive, and discard the blocks we don't need anymore. The blocks
+    // have an always-incrementing index. The index of the first block is in
+    // `first_block_number`.
+    std::deque<WindowBlock> window_blocks;
+    UInt64 first_block_number = 0;
+    // The next block we are going to pass to the consumer.
+    UInt64 next_output_block_number = 0;
+    // The first row for which we still haven't calculated the window functions.
+    // Used to determine which resulting blocks we can pass to the consumer.
+    RowNumber first_not_ready_row;
+
+    // Boundaries of the current partition.
+    // partition_start doesn't point to a valid block, because we want to drop
+    // the blocks early to save memory. We still have to track it so that we can
+    // cut off a PRECEDING frame at the partition start.
+    // The `partition_end` is past-the-end, as usual. When
+    // partition_ended = false, it still haven't ended, and partition_end is the
+    // next row to check.
+    RowNumber partition_start;
+    RowNumber partition_end;
+    bool partition_ended = false;
+
+    // The row for which we are now computing the window functions.
+    RowNumber current_row;
+    // The start of current peer group, needed for CURRENT ROW frame start.
+    // For ROWS frame, always equal to the current row, and for RANGE and GROUP
+    // frames may be earlier.
+    RowNumber peer_group_start;
+
+    // Row and group numbers in partition for calculating rank() and friends.
+    UInt64 current_row_number = 1;
+    UInt64 peer_group_start_row_number = 1;
+    UInt64 peer_group_number = 1;
+
+    // The frame is [frame_start, frame_end) if frame_ended && frame_started,
+    // and unknown otherwise. Note that when we move to the next row, both the
+    // frame_start and the frame_end may jump forward by an unknown amount of
+    // blocks, e.g. if we use a RANGE frame. This means that sometimes we don't
+    // know neither frame_end nor frame_start.
+    // We update the states of the window functions after we find the final frame
+    // boundaries.
+    // After we have found the final boundaries of the frame, we can immediately
+    // output the result for the current row, w/o waiting for more data.
+    RowNumber frame_start;
+    RowNumber frame_end;
+    bool frame_ended = false;
+    bool frame_started = false;
+
+    // The previous frame boundaries that correspond to the current state of the
+    // aggregate function. We use them to determine how to update the aggregation
+    // state after we find the new frame.
+    RowNumber prev_frame_start;
+
+    bool only_have_row_number = false;
+    bool only_have_pure_window = false;
```

Comment:
It seems that these two parameters can be used as template parameters.
Because their values are judged in the loop, there may be some performance impact.
Or add a todo to improve later.

Reply:
It is possible that the cost of conditional judgment is higher than that of addition and subtraction

Reply:
cpu pipeline prediction failure

Reply:
```
<typename WindowFunctionMode>

WindowFunctionMode::xxx
```

Reply:
> 

add todo